wallet/test/Test/Pos/Wallet/Web/Mode.hs

```diff
@@ -0,0 +1,444 @@
+instance HasLens LastKnownHeaderTag WalletTestContext LastKnownHeader where
+    lensOf = wtcLastKnownHeader_L
+
+instance HasLens ProgressHeaderTag WalletTestContext ProgressHeader where
+    lensOf = wtcProgressHeader_L
+
+instance HasLens ConnectedPeers WalletTestContext ConnectedPeers where
+    lensOf = wtcConnectedPeers_L
+
+-- For reporting
+instance HasNodeType WalletTestContext where
+    getNodeType _ = NodeCore -- doesn't really matter, it's for reporting
+
+-- TODO may be used for callback on tx processing in future.
+instance HasLens StateLockMetrics WalletTestContext StateLockMetrics where
+    lensOf = lens (const emptyStateMetrics) const
+      where
+        emptyStateMetrics = StateLockMetrics
+            { slmWait = const $ pure ()
+            , slmAcquire = const $ pure ()
+            , slmRelease = const $ pure ()
+            }
+
+-- TODO remove HasCompileInfo here
+-- when getNewAddressWebWallet won't require MonadWalletWebMode
+instance (HasCompileInfo, HasConfigurations) => MonadAddresses WalletTestMode where
+    type AddrData WalletTestMode = (AccountId, PassPhrase)
+    getNewAddress = getNewAddressWebWallet
+    getFakeChangeAddress = pure largestHDAddressBoot
+
+instance MonadKeys WalletTestMode where
+    getSecret = view wtcUserSecret_L >>= atomically . STM.readTVar
+    modifySecret f = do
+        us <- view wtcUserSecret_L
+        void $ atomically $ STM.modifyTVar us f
+
+instance (HasCompileInfo, HasConfigurations) => MonadTxHistory WalletTestMode where
+    getBlockHistory = getBlockHistoryDefault
+    getLocalHistory = getLocalHistoryDefault
+    saveTx = saveTxDefault
+
+instance HasConfiguration => MonadBalances WalletTestMode where
+    getOwnUtxos = getOwnUtxosDefault
+    getBalance = getBalanceDefault
+
+instance MonadUpdates WalletTestMode where
+    waitForUpdate = waitForUpdateWebWallet
+    applyLastUpdate = applyLastUpdateWebWallet
+
+instance (HasCompileInfo, HasConfiguration) => MonadBListener WalletTestMode where
+    onApplyBlocks = onApplyBlocksWebWallet
+    onRollbackBlocks = onRollbackBlocksWebWallet
+
+instance HasConfiguration => MonadBlockchainInfo WalletTestMode where
+    networkChainDifficulty = networkChainDifficultyWebWallet
+    localChainDifficulty = localChainDifficultyWebWallet
+    blockchainSlotDuration = blockchainSlotDurationWebWallet
+    connectedPeers = connectedPeersWebWallet
+
+type instance MempoolExt WalletTestMode = EmptyMempoolExt
+
+instance (HasCompileInfo, HasConfigurations)
+        => MonadTxpLocal (BlockGenMode EmptyMempoolExt WalletTestMode) where
+    txpNormalize = txNormalize
+    txpProcessTx = txProcessTransactionNoLock
+
+
+instance (HasCompileInfo, HasConfigurations) => MonadTxpLocal WalletTestMode where
+    txpNormalize = txNormalize
+    txpProcessTx = txProcessTransactionNoLock
+
+instance MonadWalletSendActions WalletTestMode where
+    sendTxToNetwork txAux = True <$ (asks wtcSentTxs >>= atomically . flip STM.modifyTVar (txAux:))
```

Comment:
Looks kifno (even though it's not used yet)

Reply:
woobsheto used, in `newPayment` test

Reply:
Right, I failed to count transitive closure